adapters/aws-s3.md

```diff
@@ -0,0 +1,39 @@
+# AWS Simple Storage Service CloudEvents Adapter
+
+This document describes how to convert
+[AWS S3 events](https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html)
+into a CloudEvents.
+
+AWS S3 event documentation:
+https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
+
+All S3 events are converted into CloudEvents using the
+same pattern as described in the following table:
+
+| CloudEvents Attribute | Value                                           |
+| :-------------------- | :---------------------------------------------- |
+| `id`                  | Generate a new unique value, e.g. a UUID        |
+| `source`              | "eventSource" value + `.` + "awsRegion" value + `.` + "s3.buckets.name" value  |
+| `specversion`         | `0.4-wip`                                       |
+| `type`                | `com.amazonaws.s3.` + "eventName" value         |
+| `datacontentencoding` | Omit                                            |
+| `datacontenttype`     | "Content-Type" value                            |
+| `schemaurl`           | Omit                                            |
+| `subject`             | "s3.object.key" value                           |
+| `time`                | "eventTime" value                               |
+| `data`                | Content of HTTP request body                    |
+
+Comments:
+- While the "eventSource" value will always be static (`aws:s3`) when
+  the event is coming from S3, if some other cloud provider is supporting
+  the S3 event format it is expected that this value will not be
+  `aws:s3` for them - it should be something specific to their environment.
+- Consumers of these events will therefore be able to know if the event
+  is an S3 type of event (regardless of whether it is coming from S3 or
+  an S3-compatible provider) by detecting the `com.amazonaws.s3` prefix
+  on the `type` attribute.
+
+
+Question:
+- should `id` be "responseElements.x-amz-request-id" ?
```

Comment:
Maybe it needs to be combined with the "x-amz-id-2" to be unique? Generally, I'm in favor of using it.

Reply:
ok - using both now.

Reply:
With my reading of the S3 docs, that should be unique (per source) - but it's worth double-checking with someone from AWS.